### PR TITLE
uiomove_object nocap

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

**Comment by jrtc27** [comment]
Hm, I wonder why in https://github.com/CTSRD-CHERI/cheribsd/pull/1047/files#diff-9bd357e75c93293554ec7b9d3817e70aae08dd607ff978fe902d21e7b521cb31 we thought the shm case was worth preserving

**Review by brooksdavis — APPROVED** [review]
I wonder if it make sense to add a test for tmp contingent on /tmp being tmpfs (the default in qemu)?

**Comment by bsdjhb** [comment]
> Hm, I wonder why in https://github.com/CTSRD-CHERI/cheribsd/pull/1047/files#diff-9bd357e75c93293554ec7b9d3817e70aae08dd607ff978fe902d21e7b521cb31 we thought the shm case was worth preserving

I suspect we were focused on disabling caps for the other use cases at the time.  I think it also wasn't obvious that this effectively bypassed the OBJ_HASCAP check as it's location in uipc_shm.c probably made it seem as if it was only used for shm_open and so "safe".  I think though talking with Brooks today that even on shm_open() there's no real reason to permit read(2) and write(2) of capabilities.

The one odd case we still have left is the one use case for uiomove_cap() in the CAM target driver which sends CCBs to/from the kernel via read() and write() (rather than an ioctl()).  Possibly that is a bad idea and it should just be using ioctls instead.

**Comment by brooksdavis** [comment]
> The one odd case we still have left is the one use case for uiomove_cap() in the CAM target driver which sends CCBs to/from the kernel via read() and write() (rather than an ioctl()). Possibly that is a bad idea and it should just be using ioctls instead.

The odds the read code is leaking kernel capabilities to userspace is higher than I'd like...

**Comment by bsdjhb** [comment]
> > The one odd case we still have left is the one use case for uiomove_cap() in the CAM target driver which sends CCBs to/from the kernel via read() and write() (rather than an ioctl()). Possibly that is a bad idea and it should just be using ioctls instead.
> 
> The odds the read code is leaking kernel capabilities to userspace is higher than I'd like...

I had misremembered.  It is not copying wholesale `union ccb` objects.  Instead, it is returning a user pointer (similar to the `void *udata` member in `struct kevent`).  That is, to submit an I/O request, you construct `union ccb` in userland and then write(2) a pointer to it to the target driver.  Then you can kevent/poll on the target device to wait for an I/O request to complete (you can have multiple in flight).  Once you do you read(2) on the device to get back the pointer to a completed request.  It's not too dissimilar from how POSIX AIO works except for abusing read/write to pass the pointers around instead of dedicated system calls.

**Comment by brooksdavis** [comment]
> > > The one odd case we still have left is the one use case for uiomove_cap() in the CAM target driver which sends CCBs to/from the kernel via read() and write() (rather than an ioctl()). Possibly that is a bad idea and it should just be using ioctls instead.
> > ...
> 
> I had misremembered. It is not copying wholesale `union ccb` objects. Instead, it is returning a user pointer (similar to the `void *udata` member in `struct kevent`). That is, to submit an I/O request, you construct `union ccb` in userland and then write(2) a pointer to it to the target driver. Then you can kevent/poll on the target device to wait for an I/O request to complete (you can have multiple in flight). Once you do you read(2) on the device to get back the pointer to a completed request. It's not too dissimilar from how POSIX AIO works except for abusing read/write to pass the pointers around instead of dedicated system calls.

I wonder if we want an open(2) flag to allow this.

**Comment by brooksdavis** [comment]
> > > The one odd case we still have left is the one use case for uiomove_cap() in the CAM target driver which sends CCBs to/from the kernel via read() and write() (rather than an ioctl()). Possibly that is a bad idea and it should just be using ioctls instead.
> > ...
> 
> I had misremembered. It is not copying wholesale `union ccb` objects. Instead, it is returning a user pointer (similar to the `void *udata` member in `struct kevent`). That is, to submit an I/O request, you construct `union ccb` in userland and then write(2) a pointer to it to the target driver. Then you can kevent/poll on the target device to wait for an I/O request to complete (you can have multiple in flight). Once you do you read(2) on the device to get back the pointer to a completed request. It's not too dissimilar from how POSIX AIO works except for abusing read/write to pass the pointers around instead of dedicated system calls.

I wonder if we want an open(2) flag to allow this.

**Comment by bsdjhb** [comment]
I don't know that we need an open(2) flag as the one use case is already handled in the specific character device driver.

File: bin/cheribsdtest/cheribsdtest_fs.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	if (strcmp(sb.f_fstypename, "tmpfs") == 0)
+		return (NULL);
+	else
+		return ("/tmp is not using tmpfs");
+}
+
+static int
+create_tempfile()
```
Comment (jrtc27) [inline bin/cheribsdtest/cheribsdtest_fs.c:61]: (void)

Comment (bsdjhb) [reply bin/cheribsdtest/cheribsdtest_fs.c:61]: Oh, copied from the other version of this function, but I'll fix.

One question is if the tests in cheribsdtest_vm.c that create files in /tmp should really be in _fs.c instead.

**Pull request merged**